### ISSUE TITLE
make pxf failed

### ISSUE THREAD

**Issue by zhaoyuan2022** [greenplum-db/pxf#841]
Hi,

I have an Ubuntu dev env with gpdb7 running on it. Then I  cloned the pxf code into the contrib folder under gpdb source code path. when  I was trying to build pxf, I came across these errors:

gpadmin@dev-50:~/workspace/gpdb/contrib/pxf/external-table$ make
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-but-set-variable -Werror=implicit-fallthrough=3 -Wno-format-truncation -Wno-stringop-truncation -g -O0 -g3 -ggdb  -Werror=uninitialized -Werror=implicit-function-declaration -fPIC -DPXF_API_VERSION=\"16\" -I. -I./ -I/usr/local/pdb/include/postgresql/server -I/usr/local/pdb/include/postgresql/internal -I /usr/include/foundationdb/   -D_GNU_SOURCE -I/usr/include/libxml2   -c -o src/pxfprotocol.o src/pxfprotocol.c
In file included from src/pxfbridge.h:25,
                 from src/pxfprotocol.c:21:
src/pxfuriparser.h:55:2: error: unknown type name ‘List’
   55 |  List    *options;  /* list of OptionData       */
      |  ^~~~
src/pxfuriparser.h:69:55: error: unknown type name ‘List’
   69 | void  GPHDUri_verify_core_options_exist(GPHDUri *uri, List *coreOptions);
      |                                                       ^~~~
src/pxfprotocol.c:24:10: fatal error: access/fileam.h: No such file or directory
   24 | #include "access/fileam.h"
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [<builtin>: src/pxfprotocol.o] Error 1

I can't generated the pxf.so package, then the pxf extension can not be used.
Is it an known issue, or do you have workaround for it?

Thanks!

**Comment by bradfordb-vmware** [comment]
There are changes in the GP7 branch that move the location of headers needed for building an external table extension (i.e., the pxf.so). For GP7, our current focus is on using the newer FDW-based extension.

**Comment by zhaoyuan2022** [comment]
> There are changes in the GP7 branch that move the location of headers needed for building an external table extension (i.e., the pxf.so). For GP7, our current focus is on using the newer FDW-based extension.

I can not make pxf.so successfully,  only pxfutils.o and pxffilters.o are generated under external-tables. But when I tried to create the external table, the location depends on the pxf protocol , this results in the pxf_fdw extension unavailable .

**Comment by zhaoyuan2022** [comment]
Make successfully in gpdb6 env, and failed in gpdb7 env(only different db version), the failed message of gpdb7 env is:

'
src/pxfuriparser.c: In function ‘GPHDUri_free_options’:
src/pxfuriparser.c:27:16: error: implicit declaration of function ‘list_head’ [-Werror=implicit-function-declaration]
   27 |  for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
      |                ^~~~~~~~~
src/pxfuriparser.c:251:2: note: in expansion of macro ‘foreach’
  251 |  foreach(option, uri->options) {
      |  ^~~~~~~
src/pxfuriparser.c:27:14: warning: assignment to ‘ListCell *’ {aka ‘struct ListCell *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   27 |  for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
      |              ^
src/pxfuriparser.c:251:2: note: in expansion of macro ‘foreach’
  251 |  foreach(option, uri->options) {
      |  ^~~~~~~
In file included from src/pxfuriparser.c:20:
src/pxfuriparser.h:34:27: error: dereferencing pointer to incomplete type ‘ListCell’ {aka ‘struct ListCell’}
   34 | #define lnext(lc)    ((lc)->next)
      |                           ^~
src/pxfuriparser.c:27:55: note: in expansion of macro ‘lnext’
   27 |  for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
      |                                                       ^~~~~
src/pxfuriparser.c:251:2: note: in expansion of macro ‘foreach’
  251 |  foreach(option, uri->options) {
      |  ^~~~~~~
src/pxfuriparser.c:252:37: error: implicit declaration of function ‘lfirst’ [-Werror=implicit-function-declaration]
  252 |   OptionData *data = (OptionData *) lfirst(option);
      |                                     ^~~~~~
src/pxfuriparser.c:252:22: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  252 |   OptionData *data = (OptionData *) lfirst(option);
      |                      ^
src/pxfuriparser.c:258:2: error: implicit declaration of function ‘list_free’ [-Werror=implicit-function-declaration]
  258 |  list_free(uri->options);
      |  ^~~~~~~~~
src/pxfuriparser.c: In function ‘GPHDUri_opt_exists’:
src/pxfuriparser.c:27:14: warning: assignment to ‘ListCell *’ {aka ‘struct ListCell *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   27 |  for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
      |              ^
src/pxfuriparser.c:272:2: note: in expansion of macro ‘foreach’
  272 |  foreach(item, uri->options)
      |  ^~~~~~~
src/pxfuriparser.c:274:22: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  274 |   OptionData *data = (OptionData *) lfirst(item);
      |                      ^
src/pxfuriparser.c: In function ‘GPHDUri_verify_no_duplicate_options’:
src/pxfuriparser.c:27:14: warning: assignment to ‘ListCell *’ {aka ‘struct ListCell *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   27 |  for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
      |              ^
src/pxfuriparser.c:293:2: note: in expansion of macro ‘foreach’
  293 |  foreach(option, uri->options)
      |  ^~~~~~~
src/pxfuriparser.c:295:22: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  295 |   OptionData *data = (OptionData *) lfirst(option);
      |                      ^
src/pxfuriparser.c:297:3: error: unknown type name ‘Value’
  297 |   Value    *key = makeString(asc_toupper(data->key, strlen(data->key)));
      |   ^~~~~
src/pxfuriparser.c:297:19: error: implicit declaration of function ‘makeString’ [-Werror=implicit-function-declaration]
  297 |   Value    *key = makeString(asc_toupper(data->key, strlen(data->key)));
      |                   ^~~~~~~~~~
src/pxfuriparser.c:297:19: warning: initialization of ‘int *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
src/pxfuriparser.c:302:8: error: implicit declaration of function ‘list_member’ [-Werror=implicit-function-declaration]
  302 |   if (!list_member(previousKeys, key))
      |        ^~~~~~~~~~~
src/pxfuriparser.c:308:36: error: dereferencing pointer to incomplete type ‘List’ {aka ‘struct List’}
  308 |  if (duplicateKeys && duplicateKeys->length > 0)
      |                                    ^~
src/pxfuriparser.c:311:3: error: unknown type name ‘StringInfoData’; use ‘struct’ keyword to refer to the type
  311 |   StringInfoData duplicates;
      |   ^~~~~~~~~~~~~~
      |   struct
src/pxfuriparser.c:314:3: error: implicit declaration of function ‘initStringInfo’ [-Werror=implicit-function-declaration]
  314 |   initStringInfo(&duplicates);
      |   ^~~~~~~~~~~~~~
src/pxfuriparser.c:27:14: warning: assignment to ‘ListCell *’ {aka ‘struct ListCell *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   27 |  for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
      |              ^
src/pxfuriparser.c:315:3: note: in expansion of macro ‘foreach’
  315 |   foreach(key, duplicateKeys)
      |   ^~~~~~~
src/pxfuriparser.c:317:23: error: implicit declaration of function ‘strVal’; did you mean ‘strcat’? [-Werror=implicit-function-declaration]
  317 |    char    *keyname = strVal((Value *) lfirst(key));
      |                       ^~~~~~
      |                       strcat
src/pxfuriparser.c:317:31: error: ‘Value’ undeclared (first use in this function); did you mean ‘T_Value’?
  317 |    char    *keyname = strVal((Value *) lfirst(key));
      |                               ^~~~~
      |                               T_Value
src/pxfuriparser.c:317:31: note: each undeclared identifier is reported only once for each function it appears in
src/pxfuriparser.c:317:38: error: expected expression before ‘)’ token
  317 |    char    *keyname = strVal((Value *) lfirst(key));
      |                                      ^
src/pxfuriparser.c:320:5: error: implicit declaration of function ‘appendStringInfoString’ [-Werror=implicit-function-declaration]
  320 |     appendStringInfoString(&duplicates, ", ");
      |     ^~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/gpdb/include/postgresql/server/postgres.h:47,
                 from src/pxfuriparser.h:23,
                 from src/pxfuriparser.c:20:
src/pxfuriparser.c:327:76: error: request for member ‘data’ in something not a structure or union
  327 |      errmsg("invalid URI %s: Duplicate option(s): %s", uri->uri, duplicates.data)));
      |                                                                            ^
/usr/local/gpdb/include/postgresql/server/utils/elog.h:155:4: note: in definition of macro ‘ereport_domain’
  155 |    __VA_ARGS__, errfinish(__FILE__, __LINE__, PG_FUNCNAME_MACRO); \
      |    ^~~~~~~~~~~
src/pxfuriparser.c:325:3: note: in expansion of macro ‘ereport’
  325 |   ereport(ERROR,
      |   ^~~~~~~
/usr/local/gpdb/include/postgresql/server/utils/elog.h:155:15: warning: left-hand operand of comma expression has no effect [-Wunused-value]
  155 |    __VA_ARGS__, errfinish(__FILE__, __LINE__, PG_FUNCNAME_MACRO); \
      |               ^
/usr/local/gpdb/include/postgresql/server/utils/elog.h:172:2: note: in expansion of macro ‘ereport_domain’
  172 |  ereport_domain(elevel, TEXTDOMAIN, __VA_ARGS__)
      |  ^~~~~~~~~~~~~~
src/pxfuriparser.c:325:3: note: in expansion of macro ‘ereport’
  325 |   ereport(ERROR,
      |   ^~~~~~~
src/pxfuriparser.c:328:19: error: request for member ‘data’ in something not a structure or union
  328 |   pfree(duplicates.data);
      |                   ^
src/pxfuriparser.c:330:2: error: implicit declaration of function ‘list_free_deep’ [-Werror=implicit-function-declaration]
  330 |  list_free_deep(duplicateKeys);
      |  ^~~~~~~~~~~~~~
src/pxfuriparser.c: In function ‘GPHDUri_verify_core_options_exist’:
src/pxfuriparser.c:342:2: error: unknown type name ‘StringInfoData’; use ‘struct’ keyword to refer to the type
  342 |  StringInfoData missing;
      |  ^~~~~~~~~~~~~~
      |  struct
src/pxfuriparser.c:27:14: warning: assignment to ‘ListCell *’ {aka ‘struct ListCell *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   27 |  for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
      |              ^
src/pxfuriparser.c:347:2: note: in expansion of macro ‘foreach’
  347 |  foreach(coreOption, coreOptions)
      |  ^~~~~~~
src/pxfuriparser.c:349:28: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  349 |   char    *coreOptionStr = (char *) lfirst(coreOption);
      |                            ^
src/pxfuriparser.c:27:14: warning: assignment to ‘ListCell *’ {aka ‘struct ListCell *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   27 |  for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
      |              ^
src/pxfuriparser.c:353:3: note: in expansion of macro ‘foreach’
  353 |   foreach(option, uri->options)
      |   ^~~~~~~
src/pxfuriparser.c:355:20: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  355 |    char    *key = ((OptionData *) lfirst(option))->key;
      |                    ^
In file included from /usr/local/gpdb/include/postgresql/server/postgres.h:47,
                 from src/pxfuriparser.h:23,
                 from src/pxfuriparser.c:20:
src/pxfuriparser.c:375:70: error: request for member ‘data’ in something not a structure or union
  375 |      errmsg("invalid URI %s: %s option(s) missing", uri->uri, missing.data)));
      |                                                                      ^
/usr/local/gpdb/include/postgresql/server/utils/elog.h:155:4: note: in definition of macro ‘ereport_domain’
  155 |    __VA_ARGS__, errfinish(__FILE__, __LINE__, PG_FUNCNAME_MACRO); \
      |    ^~~~~~~~~~~
src/pxfuriparser.c:373:3: note: in expansion of macro ‘ereport’
  373 |   ereport(ERROR,
      |   ^~~~~~~
/usr/local/gpdb/include/postgresql/server/utils/elog.h:155:15: warning: left-hand operand of comma expression has no effect [-Wunused-value]
  155 |    __VA_ARGS__, errfinish(__FILE__, __LINE__, PG_FUNCNAME_MACRO); \
      |               ^
/usr/local/gpdb/include/postgresql/server/utils/elog.h:172:2: note: in expansion of macro ‘ereport_domain’
  172 |  ereport_domain(elevel, TEXTDOMAIN, __VA_ARGS__)
      |  ^~~~~~~~~~~~~~
src/pxfuriparser.c:373:3: note: in expansion of macro ‘ereport’
  373 |   ereport(ERROR,
      |   ^~~~~~~
src/pxfuriparser.c:376:15: error: request for member ‘data’ in something not a structure or union
  376 |  pfree(missing.data);
      |               ^
cc1: some warnings being treated as errors
make: *** [<builtin>: src/pxfuriparser.o] Error 1
'

**Comment by zhaoyuan2022** [comment]
Do you have a link that I can download a pxf_fdw binary for gp7 on ubunbu?

**Comment by bradfordb-vmware** [comment]
We don't yet have a public build of PXF for GPDB7. Can you try running `make` inside of the `fdw` directory? That should build the PXF FDW extension against GPDB7 without error.

**Comment by zhaoyuan2022** [comment]
> We don't yet have a public build of PXF for GPDB7. Can you try running `make` inside of the `fdw` directory? That should build the PXF FDW extension against GPDB7 without error.

Thank you for the reply.  I can build pxf_fdw and install it successfully. 
But when I tried to create external table with:
`
CREATE EXTERNAL TABLE pxf_read_test (a TEXT, b TEXT, c TEXT) 
LOCATION ('pxf://localhost:5888/tmp/dummy1' 
'?FRAGMENTER=org.greenplum.pxf.api.examples.DemoFragmenter' 
'&ACCESSOR=org.greenplum.pxf.api.examples.DemoAccessor' 
'&RESOLVER=org.greenplum.pxf.api.examples.DemoTextResolver') 
FORMAT 'TEXT' (DELIMITER ',');
`

I got the error said that `ERROR:  protocol "s3" does not exist`, and I tried other protocol, such as `s3` and `hdfs`, failed either. Am I using it in a wrong way? Or the required protocol is not built and installed?
